### PR TITLE
Remove test-only code UpdateServiceShards from production codepath

### PR DESCRIPTION
**Pull request by howardjohn**
Builds on https://github.com/istio/istio/pull/42687

### CODE REVIEW

File: pilot/pkg/xds/discovery.go [hunk 1]
```diff
@@ -543,10 +518,6 @@ func (s *DiscoveryServer) initPushContext(req *model.PushRequest, oldPushContext
 		return nil, err
 	}
 
-	if err := s.UpdateServiceShards(push); err != nil {
```
Comment (hzxuzhonghu) [inline pilot/pkg/xds/discovery.go:546]: Why remove this and `getNonK8sRegistries`. 

It was originally intended for ADSC

Comment (hzxuzhonghu) [reply pilot/pkg/xds/discovery.go:546]: Oh `MCP` registry was removed, that was long ago

Comment (howardjohn) [reply pilot/pkg/xds/discovery.go:546]: It was only used for the mock client in tests currently

File: pilot/pkg/xds/eds_test.go [hunk 2]
```diff
@@ -602,7 +603,7 @@ func TestEndpointFlipFlops(t *testing.T) {
 func TestDeleteService(t *testing.T) {
 	s := xds.NewFakeDiscoveryServer(t, xds.FakeOptions{})
 	addEdsCluster(s, "removeservice.com", "http", "10.0.0.53", 8080)
-
+	reconcileServiceShards(s, s.MemServiceRegistry)
```
Comment (ramaraochavali) [inline pilot/pkg/xds/eds_test.go:606]: Do we need this here? We are already doing inside addEdsCluster()?

**Review by ramaraochavali — COMMENTED** [review]
Should MemRegistry have a reconcilerFunc that it calls when endpoints are set and instances are added so that we do not have to worry about calling that function across all tests?

**Comment by howardjohn** [comment]
> Should MemRegistry have a reconcilerFunc that it calls when endpoints are set and instances are added so that we do not have to worry about calling that function across all tests?

Yes, that was my hope but I didn't have time to make it work properly. I was focusing on moving our hacky code into tests instead of prod first

**Review by ramaraochavali — APPROVED** [review]

**Pull request merged**